### PR TITLE
[contain-intrinsic-size] Element removed by parent should end up losing last remembered size

### PR DESCRIPTION
**Pull request by cathiechen**
#### a08d2b5f36279adc3dc9b15ad42a132fe5ce065e
<pre>
[contain-intrinsic-size] Element removed by parent should end up losing last remembered size
<a href="https://bugs.webkit.org/show_bug.cgi?id=270240">https://bugs.webkit.org/show_bug.cgi?id=270240</a>

Reviewed by NOBODY (OOPS!).

The code to clear last remembered size for disconnected element, it does not cover all the scenario.
This patch moves it to Element::removedFromAncestor, so that it could handle all scenarios with the element removed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-006.html:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::removeChild):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::removedFromAncestor):
</pre>
https://github.com/WebKit/WebKit/commit/a08d2b5f36279adc3dc9b15ad42a132fe5ce065e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21174 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44550 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44747 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38272 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/44463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24372 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18505 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42730 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/18113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/42027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/37338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46190 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/35612 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/37666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/41784 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16972 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/42204 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18591 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/36608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/48793 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18652 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/9913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18236 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/dom/ContainerNode.cpp [hunk 1]
```diff
@@ -773,6 +765,14 @@ void ContainerNode::removeBetween(Node* previousChild, Node* nextChild, Node& ol
     oldChild.setParentNode(nullptr);
 
     oldChild.setTreeScopeRecursively(document());
+
+    auto* element = dynamicDowncast<Element>(oldChild);
```
Comment (rniwa) [inline Source/WebCore/dom/ContainerNode.cpp]: Can we add this to Element::removedFromAncestor instead?
We can check `!parentNode()` to detect the top-most node in a removed subtree.

Comment (cathiechen) [reply Source/WebCore/dom/ContainerNode.cpp]: `ContainerNode::removeBetween` is called `ContainerNode::removeAllChildrenWithScriptAssertion` and `ContainerNode::removeNodeWithScriptAssertion` which are triggered by JS. While  `Element::removedFromAncestor` besides the previous two functions, it is also called by `ContainerNode::removeDetachedChildren`, which is triggered by GC, technically, in this scenario we don't need to handle the element if it is already garbage collected. So it seems more efficient if we keep the changes here.

Regarding the `!parentNode()` part, I don't quite understand it, could you explain more? Thanks!

Comment (rniwa) [reply Source/WebCore/dom/ContainerNode.cpp]: Yeah but generally, removedFromAncestor is where we update these kinds of caches so it's more appropriate for it to be there. Besides that, removeBetween is a core DOM mutation helper so logic specific to IntersectionObserver doesn't really belong there.

My commentary about `!parentNode()` is that `removedFromAncestor` gets called on a node when it was removed from its immediate parent, or its ancestor is removed from its parent. So in order to run the code only when its immediate parent is changed, we need to check `!parentNode()` so your code in `Element::removedFromAncestor` will look like this:
```
if (!parentNode() && (lastRememberedLogicalWidth() || lastRememberedLogicalHeight())) {
    document().observeForContainIntrinsicSize(*this);
    document().resetObservationSizeForContainIntrinsicSize(*this);
}
```

Comment (cathiechen) [reply Source/WebCore/dom/ContainerNode.cpp]: @rniwa I see, thanks for the explanation!

**Pull request closed without merging**